polygeist/tools/cgeist/Lib/CGCall.cc

```diff
@@ -1528,6 +1529,30 @@ ValueCategory MLIRScanner::VisitCallExpr(clang::CallExpr *expr) {
   MLIRScanner::getMangledFuncName(name, callee, Glob.getCGM());
   if (isSupportedFunctions(name))
     ShouldEmit = true;
+
+  auto findCaller = [&](clang::CallExpr *expr) {
+    clang::DynTypedNodeList nodeList =
+        Glob.getCGM().getContext().getParents(*expr);
+    while (!nodeList.empty()) {
+      clang::DynTypedNode parentNode = nodeList[0];
+
+      if (const FunctionDecl *parent = parentNode.get<FunctionDecl>())
+        return parent;
+
+      nodeList = Glob.getCGM().getContext().getParents(parentNode);
+    }
+
+    llvm_unreachable("cannot find caller");
+    return (const FunctionDecl *)nullptr;
+  };
+  const FunctionDecl *caller = findCaller(expr);
+  bool isCallerDevice = caller->hasAttr<clang::SYCLKernelAttr>() ||
+                        caller->hasAttr<clang::SYCLDeviceAttr>();
+  if (isCallerDevice) {
+    LLVM_DEBUG(llvm::dbgs() << "Adding device attribute to " << name << "\n");
+    const_cast<FunctionDecl *>(callee)->addAttr(
+        SYCLDeviceAttr::CreateImplicit(Glob.getCGM().getContext()));
+  }
```

Comment:
It will be nice to have some comments in the above blob of code to indicate what/why they are doing